### PR TITLE
Testcafe screenshot test for pager

### PR DESCRIPTION
**Pull request by MikeVitik**

### CODE REVIEW

**Comment by San4es** [comment]
Let's rename workflow to "Tests", "CI" or smth else:
https://github.com/DevExpress/DevExtreme/actions/runs/337844025/workflow#L1

**Review by San4es — DISMISSED** [review]

**Review by San4es — DISMISSED** [review]

**Review by NickMitrokhin — DISMISSED** [review]

File: testing/testcafe/helpers/screenshort-comparer.ts [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+  function copyToArtifacts(sourcePath: string, postfix = '') {
+    const fileName = path.basename(sourcePath, '.png');
+    const targetPath = path.join(artifactsPath, `${fileName}${postfix}.png`);
+    if (fs.existsSync(sourcePath)) {
+      fs.copyFileSync(sourcePath, targetPath);
+    }
+  }
+
+  copyToArtifacts(screenshotFileName);
+  copyToArtifacts(etalonFileName, '_ethalon');
```
Comment (viterobk) [inline testing/testcafe/helpers/screenshort-comparer.ts]: _ethalon -> _etalon

File: testing/testcafe/helpers/screenshort-comparer.ts [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+  });
+}
+
+async function getMaskedScreenshotBuffer({
+  screenshotFileName, etalonFileName, maskFileName,
+}) {
+  function isSizeEqual(image1, image2): boolean {
+    return image1.height === image2.height && image1.width === image2.width;
+  }
+  function getImage(imagePath: string): { width; height} | undefined {
```
Comment (viterobk) [inline testing/testcafe/helpers/screenshort-comparer.ts]: `getImage` function's return type looks incomplete

File: testing/testcafe/helpers/screenshort-comparer.ts [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    return PNG.sync.write(screenshotImg);
+  }
+
+  const etalonImg = getImage(etalonFileName);
+  const screenshotImg = getImage(screenshotFileName);
+  const maskImg = getImage(maskFileName);
+
+  if (!fs.existsSync(etalonFileName)) {
+    throw new Error(`Etalon file not found: ${etalonFileName}`);
+  }
```
Comment (viterobk) [inline testing/testcafe/helpers/screenshort-comparer.ts:94]: Should not this check precede the `getImage(etalonFileName)` call?

File: testing/testcafe/helpers/screenshort-comparer.ts [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+  if (!isSizeEqual(etalonImg, maskImg)) {
+    throw new Error('Mask size does not match etalon size');
+  }
+  const targetImageBuffer = applyMask({ etalonImg, screenshotImg, maskImg });
+  return targetImageBuffer;
+}
+
+async function getDiff({
+  etalonFileName, screenshotBuffer, options,
+}: { etalonFileName; screenshotBuffer; options: ComparerOptions}):
```
Comment (viterobk) [inline testing/testcafe/helpers/screenshort-comparer.ts]: Missing space before closing curly brace

File: testing/testcafe/helpers/screenshort-comparer.ts [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+  const errorMessages: string[] = [];
+  return {
+    takeScreenshot: async (screenshotName: string,
+      element: SelectorType = null,
+      comparisonOptions?: Partial<ComparerOptions>,
+    ) => {
+      try {
+        const isValid = await compareScreenshot(t, screenshotName, element, comparisonOptions);
+        if (!isValid) {
+          errorMessages.push(`Screenshort:'${screenshotName}' isn't valid`);
```
Comment (viterobk) [inline testing/testcafe/helpers/screenshort-comparer.ts]: I'd say "Screenshort:'${screenshotName}' is invalid"
Also pay attention at the typo: Screensho>r<t

File: testing/testcafe/helpers/screenshort-comparer.ts [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+        }
+      } catch (e) {
+        errorMessages.push(`Screenshort:'${screenshotName}' isn't valid, internalError: ${e.message}`);
```
Comment (viterobk) [inline testing/testcafe/helpers/screenshort-comparer.ts]: I'd say "Screenshort:'${screenshotName}' validation failed, internal error: ${e.message}"
Also pay attention at the typo: Screensho>r<t

**Review by viterobk — APPROVED** [review]

**Review by San4es — APPROVED** [review]

**Review by NickMitrokhin — APPROVED** [review]

**Pull request merged**